channeldb/payment_status.go

```diff
@@ -161,7 +184,131 @@ func (ps PaymentStatus) Registrable(settled, paymentFailed bool) error {
 	case StatusFailed:
 		return ErrPaymentAlreadyFailed
 
+	case StatusError:
+		return ErrPaymentInternal
+
 	default:
 		return fmt.Errorf("%w: %v", ErrUnknownPaymentStatus, ps)
 	}
 }
+
+// decidePaymentStatus uses the payment's DB state to determine a memory status
+// that's used by the payment router to decide following actions.
+// Together, we use four variables to determine the payment's status,
+//   - inflight: whether there are any pending HTLCs.
+//   - settled: whether any of the HTLCs has been settled.
+//   - failed: whether any of the HTLCs has been failed.
+//   - reason: whether the payment has been marked as failed.
+//
+// Based on the above variables, we derive the status using the following
+// table,
+//
+//	| inflight | settled | failed | reason |         status       |
+//	|:--------:|:-------:|:------:|:------:|:--------------------:|
+//	|   true   |   true  |  true  |   yes  |      StatusError     |
+//	|   true   |   true  |  true  |   no   |    StatusInFlight    |
+//	|   true   |   true  |  false |   yes  |      StatusError     |
+//	|   true   |   true  |  false |   no   |    StatusInFlight    |
+//	|   true   |   false |  true  |   yes  |    StatusInFlight    |
+//	|   true   |   false |  true  |   no   |    StatusInFlight    |
+//	|   true   |   false |  false |   yes  |    StatusInFlight    |
+//	|   true   |   false |  false |   no   |    StatusInFlight    |
+//	|   false  |   true  |  true  |   yes  |      StatusError     |
```

Comment:
Is `!inflight && settled` really a reason to return status error? It seems that the recipient failed to pull all the shards, but do we care? From the perspective of the sender, the payment succeeded because we have the preimage.

Reply:
`!inflight && settled` is not the reason, here we return an error because `settled && reason != nil`.

Reply:
Same misinterpretation https://github.com/lightningnetwork/lnd/pull/7175#discussion_r1098374780